### ISSUE TITLE
How to use the library for high throughput scenarios

### ISSUE THREAD

**Issue by IndicatorFluid** [grpc/grpc-java#6703]
I have a simple RPC contract akin to:

```
rpc GetMeData(Empty) returns (stream Data)
```

I have a question how to use this library effectively for high throughput scenarios, where source of data server side is able to produce data at a higher rate than GRPC is able to dispatch the data over the network to the clients. The process that produces data uses native code, so needs to be initialized and torn down to free up resources once the RPC is over.

I was under the assumption that `onNext(msg)` would block at some point, but it seems that is not the case, and I essentially just end up running out of memory as netty tries to buffer everything.

After this realization I've moved to casting the `StreamObserver<Data>` to `ServerCallStreamObserver<Data>`, and essentially I check `stream.isReady()` before every `stream.onNext(data)`. If it's not ready I just sleep in a loop until it becomes ready. This got rid of blowing up the memory problem.

This is obviously evil, and it seems to permanently steal the threads from the executor pool, preventing other clients to be served. Also, this sort of means I have to go on an unbounded thread pool to support an unbounded number of clients.
The benefit of this is that I have a clear execution path of when the call starts, or when it finishes, so my code can tear down the process that produces the data and free resources allocated if RPC for some reason breaks.

Ton of threads that are busy sleeping is obviously not great, so I tried to move to the implementation suggested by:

https://github.com/grpc/grpc-java/issues/2247#issuecomment-245949881

However, if the client terminates prematurely, it seems my runnable just stops being called, and I have no indication that RPC call is now over and that I can clean up the resources related to the RPC call. Meaning my native code is now never freed causing me to leak resources.
Looking at the exception catching the clients going away, it seems there is very little GRPC stack frames in there, so I wonder if GRPC itself does not know if netty noticed the client going away:
```
io.netty.handler.codec.http2.Http2Exception$StreamException: Stream closed before write could take place
        at io.netty.handler.codec.http2.Http2Exception.streamError(Http2Exception.java:149) ~[netty-codec-http2-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.handler.codec.http2.DefaultHttp2RemoteFlowController$FlowState.cancel(DefaultHttp2RemoteFlowController.java:480) ~[netty-codec-http2-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.handler.codec.http2.DefaultHttp2RemoteFlowController$1.onStreamClosed(DefaultHttp2RemoteFlowController.java:104) [netty-codec-http2-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.handler.codec.http2.DefaultHttp2Connection.notifyClosed(DefaultHttp2Connection.java:356) [netty-codec-http2-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.handler.codec.http2.DefaultHttp2Connection$ActiveStreams.removeFromActiveStreams(DefaultHttp2Connection.java:1000) [netty-codec-http2-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.handler.codec.http2.DefaultHttp2Connection$ActiveStreams.deactivate(DefaultHttp2Connection.java:956) [netty-codec-http2-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.handler.codec.http2.DefaultHttp2Connection$DefaultStream.close(DefaultHttp2Connection.java:512) [netty-codec-http2-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.handler.codec.http2.DefaultHttp2Connection.close(DefaultHttp2Connection.java:152) [netty-codec-http2-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.handler.codec.http2.Http2ConnectionHandler$BaseDecoder.channelInactive(Http2ConnectionHandler.java:221) [netty-codec-http2-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.handler.codec.http2.Http2ConnectionHandler.channelInactive(Http2ConnectionHandler.java:429) [netty-codec-http2-4.1.27.Final.jar!/:4.1.27.Final]
        at io.grpc.netty.NettyServerHandler.channelInactive(NettyServerHandler.java:592) [grpc-netty-1.27.0.jar!/:1.27.0]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245) [netty-transport-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231) [netty-transport-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224) [netty-transport-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1429) [netty-transport-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245) [netty-transport-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231) [netty-transport-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:947) [netty-transport-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:822) [netty-transport-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) [netty-common-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404) [netty-common-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:464) [netty-transport-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884) [netty-common-4.1.27.Final.jar!/:4.1.27.Final]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.27.Final.jar!/:4.1.27.Final]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_141]

```

Summary of my questions: 
1. If using the async onReady approach, it seems I am never notified that the client has gone away. How can I know once the RPC call is over so I could clean up resources?
2. If I block in the RPC thread, is it expected that I will only be able to server numberOfThreadsInThePool clients at the same time? I assume if I use the onReady approach I can serve infinite clients as long as I don't block the threads?
3. Is there clear way to control how much buffering netty does, and ideally disable it? I would much prefer onNext to block if it can't put stuff directly on the network buffer, because it seems it's very hard to tame netty not to use stupid amounts of memory.
4. Is there any documentation around this? All the examples are pretty primitive without handling backpressure, and I have to crawl comments on issues to find clues of how to do this sensibly at scale without eating up all the ram.

**Comment by ericgribkoff** [comment]
Sorry for the delay in responding. The person probably best suited to answer some of these questions is OOO right now, but I can provide at least some partial responses. Please let me know if I've misunderstood part of your question or am just referring you to something you've already tried and which didn't work.

> 1. If using the async onReady approach, it seems I am never notified that the client has gone away. How can I know once the RPC call is over so I could clean up resources?

I'm unsure why your `StreamObserver` would not be notified of client closure due to using `onReady` - unless maybe your stream observer is blocking elsewhere in your implementation?

> 2. If I block in the RPC thread, is it expected that I will only be able to server numberOfThreadsInThePool clients at the same time? I assume if I use the onReady approach I can serve infinite clients as long as I don't block the threads?

Yes, I believe that is correct. There's some discussion of the threading model's interaction with the on ready handler [here](https://github.com/grpc/grpc-java/blob/bb37524750fa4923f7dc390f5f961602b46b3790/examples/src/main/java/io/grpc/examples/manualflowcontrol/ManualFlowControlServer.java#L47) in our flow control example.

> 3. Is there clear way to control how much buffering netty does, and ideally disable it? I would much prefer onNext to block if it can't put stuff directly on the network buffer, because it seems it's very hard to tame netty not to use stupid amounts of memory.

I'm not sure about knobs for this within Netty - I believe the intent here is that either gRPC or HTTP/2 level flow control (you can set the window size used by Netty via [this method](https://grpc.github.io/grpc-java/javadoc/io/grpc/netty/NettyServerBuilder.html#flowControlWindow-int-) on the server builder) handles stopping the server from buffering beyond a certain point. It sounds like you're using gRPC level flow control but perhaps not using [setOnReadyHandler](https://grpc.github.io/grpc-java/javadoc/io/grpc/stub/CallStreamObserver.html#setOnReadyHandler-java.lang.Runnable-) (see below)

> 4. Is there any documentation around this? All the examples are pretty primitive without handling backpressure, and I have to crawl comments on issues to find clues of how to do this sensibly at scale without eating up all the ram.

Have you looked at https://github.com/grpc/grpc-java/tree/master/examples/src/main/java/io/grpc/examples/manualflowcontrol? From your description of a blocking spin-and-wait for `isReady`, it sounds like you are not using the `onReadyHandler` functionality as demonstrated in the example code.

Another question that may be relevant (see [this SO question](https://stackoverflow.com/a/58875471/237399)) is if your server is handling large numbers of clients or is outpacing the network bandwidth with a smaller number of streams sending large amounts of data. My normal reference for flow control is thinking it from the perspective of not overwhelming a single stream (connection), but if you have lots and lots of unary requests coming in over different connections then there are global settings to consider as well.

**Comment by IndicatorFluid** [comment]
Hey, thanks for your response.
As you can see from the stacktrace I've posted, the path caused by "Stream closed before write could take place" does not seem to call cancellation hook, so I am not sure how I should proceed to perform cleanup.

**Comment by ericgribkoff** [comment]
> As you can see from the stacktrace I've posted, the path caused by "Stream closed before write could take place" does not seem to call cancellation hook, so I am not sure how I should proceed to perform cleanup.

I think this is in reference to question 1 in your original post. I would need some more details to answer this: specifically, where is this exception showing up? It would be helpful to see your server code to understand where you're logging this exception and why you can't clean up native resources at that point.

**Comment by IndicatorFluid** [comment]
The exception is not showing up in my code at all. It seems it gets thrown somewhere inside of netty/grpc, and I observe it just because I have logging configured.

This is a rough sketch of my Kotlin code:
```
    override fun getDataForRequest(request: Request, responseObserverNonCall: StreamObserver<Response>) {
        val nativeCodeResource = NativeCodeResourceAllocator.allocate(request)
        val responseObserver = responseObserverNonCall as ServerCallStreamObserver<Response>

        val runnable = Runnable {
            while (true) {
                if (!responseObserver.isReady) {
                    // Abort, we'll be called again by the ready handler.
                    return@Runnable
                }
                if (responseObserver.isCancelled) {
                    nativeCodeResource.free()
                    return@Runnable
                }
                
                if (nativeCodeResource.hasNext()) {
                    responseObserver.onNext(nativeCodeResource.next())
                } else {
                    responseObserver.onCompleted()
                    nativeCodeResource.free()
                    return@Runnable
                }
            }
        }

        // Tell GRPC what to call once there is space in the buffers
        responseObserver.setOnReadyHandler(runnable)
        // Tell GRPC what to call if the RPC is cancelled.
        responseObserver.setOnCancelHandler {
            nativeCodeResource.free()
        }

        // Trigger the first run by ourselves.
        runnable.run()
    }
```

It seems when connection issue described in the first post happens, my runnable is never called/woken up to be notified that the client has gone away, and the cancel is not called, so I leak the native code resources.

**Comment by ericgribkoff** [comment]
Can you post more complete logs showing where/how this exception is logged? It must be getting caught and logged somewhere, and seeing where would be helpful.

In general, I would recommend structuring your manual flow control as shown in our example [here](https://github.com/grpc/grpc-java/blob/47b6b390bb79675eaa88f511e287c2449859eebd/examples/src/main/java/io/grpc/examples/manualflowcontrol/ManualFlowControlServer.java#L37) rather than the snippet you've shown in the last comment. I'm not very familiar with Kotlin, but I believe you wanted to exert manual control over flow control which would require calling https://grpc.github.io/grpc-java/javadoc/io/grpc/stub/CallStreamObserver.html#disableAutoInboundFlowControl-- and manually requesting messages in the onReadyHandler. This doesn't AFAIK explain why you're not seeing your onCancelHandler invoked, but it should help accomplish what I think is the underlying goal here.

Related to the onCancelHandler: does this *never* get invoked? Could it be that the client has broken the stream without e.g. sending a RST_STREAM, so the server-side of the connection doesn't know the client is gone until the connection times out some time later?

**Comment by IndicatorFluid** [comment]
This is the full exception that I am seeing.
This is coming from the threadpool used by netty. I have no idea where it's logged as it just seems to come through slf4j.

The example you linked is `stream <-> stream`, in my example I have `single call -> stream`, in which case there is no flow control on the server side, as there only a single message that comes in from the client.

I am trying not to blow up netty buffers and getting OOMed by doing onNext() server side faster than netty can disaptch the data over the network, and at the same time avoid permanently stealing threads from nettys thread pool (by busy waiting for stream.isReady to become true).
It seems if you just keep doing onNext without checking isReady, you just OOM.

As far as I believe, the only time cancel handler is not invoked is when this message appears in the logs. I also have ingress controller in the way which might be abruptly terminating connections.

Obviously if client does a proper shutdown sequence and no connection errors occur, stuff works as expected.

**Comment by ericgribkoff** [comment]
The stack trace of the exception in the first post here (`io.netty.handler.codec.http2.Http2Exception$StreamException: Stream closed before write could take place`) indicates it's coming from [`NettyServerHandler#channelInactive](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyServerHandler.java#L567) which should also be going through and [cancelling](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyServerHandler.java#L581) all streams on the channel.

> As far as I believe, the only time cancel handler is not invoked is when this message appears in the logs. I also have ingress controller in the way which might be abruptly terminating connections, which leads to this exception.

How many streams do you have open? If it takes e.g. until a TCP-level timeout before Netty shuts down the channel, there could be a significant delay between the connection termination and the presence of this exception (which should trigger the `onCancelHandler`). I'm asking because in the presence of large numbers of streams with staggered abrupt terminations it may be difficult to correlate the exception you're seeing with the status of a specific stream - if you're able to observe the exception and confirm that it's not followed by a call to the cancel handler, then that seems like a bug.

cc @ejona86 As far as I can tell the server application code here looks appropriate, based on https://github.com/grpc/grpc-java/issues/2247#issuecomment-245949881. If `onCancelHandler` isn't being invoked that seems like a bug, but wanted to check with you if there may be other explanations for what's happening here.

**Comment by ejona86** [comment]
I think the problem is likely this loop:
```
                if (nativeCodeResource.hasNext()) {
                    responseObserver.onNext(nativeCodeResource.next())
```

As already mentioned, it is lacking `isReady()` checks, and yes, you will OOM yourself if you generate too much data; it has to be buffered and that takes memory. But also, while this loop is running _no other callbacks occur_. This is because the callbacks are non-thread-safe, so gRPC waits for one callback to return before calling additional callbacks.

So either the code needs to occasionally return from `onReadyHandler` or it needs to poll [`responseObserver.isCancelled()`](https://grpc.github.io/grpc-java/javadoc/io/grpc/stub/ServerCallStreamObserver.html#isCancelled--).

**Comment by IndicatorFluid** [comment]
@ejona86 that's not a loop, that is an if.
 The loop is above, and it checks both isCancelled and isReady before doing onNext.

All clients have only a single stream, but I have a few hundred clients hitting a single service (I am load testing), so potentially exactly what you are saying is happening, I just get a bunch of dead connections from clients that take `TCP-level timeout` amount of time to roll off and free up resources, hence end up OOMing. The dead connections most likely happen because of ingresses/reverse proxies falling over under load, as the service can produce the data at at hundred of megabytes a second per client.

Not really sure how to work around that other than implement liveness checks at message level?
Or does GRPC have some keepalive knobs I could tune?

**Comment by ejona86** [comment]
Doh. It totally says `if` and I copied `if`. It just seemed there had to be a loop so I invented one I guess...

It appears `while (true) {` is the loop, which still causes problems if the RPC is cancelled while blocked a long time on `hasNext()`. But it should only continue one message. And if you generate messages semi-regularly there wouldn't be any issue with the code.

So the thinking is it is taking a while to notice clients are dead? That seems fair. Keepalive is available via `NettyServerBuilder.keepAliveTime()`.
